### ISSUE TITLE
Brackets mismatched in notebook code cells 

### ISSUE THREAD

**Issue by pigpag** [jupyterlab/jupyterlab#15084]
## Description

In JupyterLab 4.0.5, with "Match Brackets" turned on for code cells, sometimes the brackets are mismatched.  The highlighted bracket does not actually match the other end, and when a newline is inserted, incorrect indentation is inserted because of this.

This issue is not present in JupyterLab 3.6.

## Reproduce

A minimum example is `(1(^1)`.  In JupyterLab 4.0.5 the closing bracket is matched against the first open bracket.

I noticed that mismatched brackets often happen when the expression is not a valid Python expression.  Nevertheless, these expressions may be valid in a magic cell or other domain-specific languages, so it is desirable to have brackets matched correctly regardless of the validity of the expression.

## Context

- Operating System and version: CenOS 7.9
- Browser and version: Chrome 110
- JupyterLab version: 4.0.5

**Comment by krassowski** [comment]
I see what you mean and can reproduce this on https://codemirror.net/try/?c=aW1wb3J0IHtiYXNpY1NldHVwLCBFZGl0b3JWaWV3fSBmcm9tICJjb2RlbWlycm9yIgppbXBvcnQge3B5dGhvbn0gZnJvbSAiQGNvZGVtaXJyb3IvbGFuZy1weXRob24iCgpuZXcgRWRpdG9yVmlldyh7CiAgZG9jOiAiKDEoXjEpXG4iLAogIGV4dGVuc2lvbnM6IFtiYXNpY1NldHVwLCBweXRob24oKV0sCiAgcGFyZW50OiBkb2N1bWVudC5ib2R5Cn0pCg==

> I noticed that mismatched brackets often happen when the expression is not a valid Python expression.

Can you find a reproducer for the case when it happens for a valid Python expression?

**Comment by pigpag** [comment]
No, I tried pretty hard to construct one but so far I haven't found any.
